pkg/controller/remotecluster/controller.go

```diff
@@ -0,0 +1,449 @@
+		return reconcile.Result{}, nil
+	}
+	return doReconcile(ctx, r, &es)
+}
+
+// deleteAllRemoteCa deletes all associated remote certificate authorities
+func deleteAllRemoteCa(ctx context.Context, r *ReconcileRemoteClusters, es types.NamespacedName) (reconcile.Result, error) {
+	span, _ := apm.StartSpan(ctx, "delete_all_remote_ca", tracing.SpanTypeApp)
+	defer span.End()
+
+	associatedCAs, err := getAssociatedRemoteCAs(ctx, r.Client, es)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	results := &reconciler.Results{}
+	for remoteCluster := range associatedCAs {
+		if err := deleteCertificateAuthorities(ctx, r, es, remoteCluster); err != nil {
+			results.WithError(err)
+		}
+	}
+	return results.Aggregate()
+}
+
+func doReconcile(
+	ctx context.Context,
+	r *ReconcileRemoteClusters,
+	localEs *esv1.Elasticsearch,
+) (reconcile.Result, error) {
+	log := ulog.FromContext(ctx)
+
+	localClusterKey := k8s.ExtractNamespacedName(localEs)
+
+	expectedRemoteClusters, err := getExpectedRemoteClusters(ctx, r.Client, localEs)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	enabled, err := r.licenseChecker.EnterpriseFeaturesEnabled(ctx)
+	if err != nil {
+		return defaultRequeue, err
+	}
+	if !enabled && len(expectedRemoteClusters) > 0 {
+		log.V(1).Info(
+			"Remote cluster controller is an enterprise feature. Enterprise features are disabled",
+			"namespace", localEs.Namespace, "es_name", localEs.Name,
+		)
+		return reconcile.Result{}, nil
+	}
+
+	// Get all the clusters to which this reconciled cluster is connected to according to the existing remote CAs.
+	// associatedRemoteCAs is used to delete the CA certificates and cancel any trust relationships
+	// that may have existed in the past but should not exist anymore.
+	associatedRemoteCAs, err := getAssociatedRemoteCAs(ctx, r.Client, localClusterKey)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	var (
+		activeAPIKeys esclient.CrossClusterAPIKeyList
+		esClient      esclient.Client
+	)
+	localClusterSupportsClusterAPIKeys, err := localEs.SupportsRemoteClusterAPIKeys()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	results := &reconciler.Results{}
+	if localClusterSupportsClusterAPIKeys.IsTrue() {
+		// Check if the ES API is available. We need it to create, update and invalidate
+		// API keys in this cluster.
+		if !services.NewElasticsearchURLProvider(*localEs, r.Client).HasEndpoints() {
+			log.Info("Elasticsearch API is not available yet")
+			return results.WithResult(defaultRequeue).Aggregate()
+		}
+		// Create a new client
+		newEsClient, err := r.esClientProvider(ctx, r.Client, r.Dialer, *localEs)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		// Check that the API is available
+		esClient = newEsClient
+		// Get all the API Keys, for that specific client, on the reconciled cluster.
+		getCrossClusterAPIKeys, err := esClient.GetCrossClusterAPIKeys(ctx, "eck-*")
```

Comment:
Should we try to handle the case where a new cluster is not reachable yet to avoid noisy error logs like this: 
```
manager.eck-operator	Reconciler error	{"service.version": "2.15.0-SNAPSHOT+65769109", "controller": "remotecluster-controller", "object": {"name":"cluster3","namespace":"ns2"}, "namespace": "ns2", "name": "cluster3", "reconcileID": "8aebea3f-118d-474e-b4c8-4b55efc26141", "error": "elasticsearch client failed for https://cluster3-es-default-0.cluster3-es-default.ns2:9200/_security/api_key?active_only=true&error_trace=true&name=eck-%2A: Get \"https://cluster3-es-default-0.cluster3-es-default.ns2:9200/_security/api_key?active_only=true&error_trace=true&name=eck-%2A\": EOF"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).reconcileHandler
	/Users/pebrc/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:316
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).processNextWorkItem
	/Users/pebrc/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:263
```

Reply:
This should be the case: https://github.com/elastic/cloud-on-k8s/blob/657691095e762897de05878900ea5a98ed9baa42/pkg/controller/remotecluster/controller.go#L181-L186

Maybe there's a bug, this `EOF` seems a bit odd though 🤔 , I would expect a connection refused or another error, `EOF` feels like the connection was initiated but unexpectedly closed.

Reply:
Could you check if this error comes from our [port-forwarder "hack"](https://github.com/elastic/cloud-on-k8s/tree/b5de7248728cd9c1b81cb71be01d44ac51bd4189/pkg/dev/portforward) used in dev mode please, maybe by looking at the full stacktrace?

Reply:
Yes that was it